AWS: Allow cross-region image pulling with ECR

therc opened the pull request:
Fixes #23298
Definitely should be in the release notes; should maybe get merged in 1.2 along with #23594 after some soaking. Documentation changes to follow.

cc @justinsb @erictune @rata @miguelfrde

This is step two. We now create long-lived, lazy ECR providers in all regions.
When first used, they will create the actual ECR providers doing the work
behind the scenes, namely talking to ECR in the region where the image lives,
rather than the one our instance is running in.

Also:
- moved the list of AWS regions out of the AWS cloudprovider and into the
  credentialprovider, then exported it from there.
- improved logging

Behold, running in us-east-1:

```
aws_credentials.go:127] Creating ecrProvider for us-west-2
aws_credentials.go:63] AWS request: ecr:GetAuthorizationToken in us-west-2
aws_credentials.go:217] Adding credentials for user AWS in us-west-2
Successfully pulled image "123456789012.dkr.ecr.us-west-2.amazonaws.com/test:latest"
```

_"One small step for a pod, one giant leap for Kube-kind."_

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/24369)

<!-- Reviewable:end -->

rata commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -59,56 +76,96 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
+// by the client when attempting to pull an image and it will create the actual
+// provider only when we actually need it the first time.
+func (p *lazyEcrProvider) LazyProvide() *credentialprovider.DockerConfigEntry {
+	if p.actualProvider == nil {
```
What do you think instead of having the code here inside the if, to create the actualProvider, move it to another function and do something like:

```
if p.actualProvider == nil {
  p.actualProvider = <func to create the actualProvider here>
}
entry := p.actualProvider.Provide()[p.regionURL]
return &entry
```

I'm not sure how the k8s code styling is (will check next week :)), but this is a small detail but it seems slightly more readable for me

therc replied:
Yes, I was going to try a helper function next. And perhaps, instead of creating N lazy providers, just one, which does a string match against the image getting loaded, figuring the region, then maintaining a map of ecrProviders.

rata replied:
I'm not sure about that. The code this way is pretty simple and the N lazy providers shouldn't make noticiable difference in mem or cpu, and as no real provider is created for the unused, not even messages with the cloud provider are used.

So, I'd do it only if it is equally or more readable than now, and equally or less error prone. If the code gets more messy, I don't think it's worth the effort.

But your call :)

therc replied:
Helper/constructor added. I'm still tempted to have one lazy provider (which would definitely require more tests).

rata commented on pkg/credentialprovider/aws/aws_credentials_test.go:
```diff
@@ -58,12 +58,15 @@ func (p *testTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationToken
 	image := "foo/bar"
 
 	provider := &ecrProvider{
+		region: "lala-land-1",
```
Just curious, is there something about the region string that is expected? Do you used a string with "-" 2 times because that is the expected format?

therc replied:
I used it because it looks a bit like a real region, but isn't one. It's a old habit, to guard against those rare times when test code ends up running unexpectedly against production.

rata commented:
@therc: nice! I did two small questions, but looks great for me (will test it next week :))

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -566,20 +566,7 @@ func getAvailabilityZone(metadata EC2Metadata) (string, error) {
-		"sa-east-1",
-	}
-	for _, r := range regions {
+	for _, r := range aws_credentials.AWSRegions {
```
Nice find!

justinsb commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -59,56 +76,96 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
 	return p.svc.GetAuthorizationToken(input)
 }
 
+// lazyEcrProvider is a DockerConfigProvider that creates on demand an
```
Tip: I personally like to do this:

```
var _ DockerConfigProvider = &lazyEcrProvider{}
```

Apparently there's some debate as to whether or not to do it, but I like that it makes the intent very clear and that it is then checked immediately by the golang compiler.

therc replied:
Added both.

justinsb commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -59,56 +76,96 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
+			&credentialprovider.CachingDockerConfigProvider{
+				Provider: &lazyEcrProvider{
+					region:    region,
+					regionURL: fmt.Sprintf(registryURLTemplate, region),
```
I do wonder if there is value to declaring this as a constant, vs just inlining it.  I find this harder to read, and the constant seems to only be used once.  Up to you.

therc replied:
Well, with the constructor, it's used twice now. :-)

justinsb commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -59,56 +76,96 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
+				},
+				// This is going to be just a lazy proxy to the real ecrProvider.
+				// It holds no real credentials, so refresh practically never.
+				Lifetime: 365 * 24 * time.Hour,
```
I'm a little confused about why this "extra" CachingDockerConfigProvider is needed (vs just using lazyEcrProvider)?  I'm sure it's just that I'm not (yet) overly familiar with this code, but I'm doubly confused by never refreshing the token...

rata replied:
Didn't checked the code in credentialprovider, but I *guess* it is not an optional argument. And is registering this lazy provider as the provider. If one is actually used, then the real provider is created.

But maybe I'm not understanding correctly. But this makes me wonder, why not do the lazyness on credentialprovider? Or we don't want the lazyness at that abstraction level? (never even looked at the code, so it might be a totally stupid question, sorry in advance)

therc replied:
You're right, @justinsb, perhaps the CachingDockerConfigProvider is unnecessary, unless we need it for type safety. I'll look into that. We never refresh the token for the lazy provider because it's just a shell for another provider (ecrProvider) that gets created on demand. There are no real credentials inside it to refresh anyway.

therc replied:
Removed the caching wrapper.

justinsb commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -59,56 +76,96 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
+
+// Enabled implements DockerConfigProvider.Enabled for the lazy provider.
+func (p *lazyEcrProvider) Enabled() bool {
+	return true
```
actualProvider.Enabled?

rata replied:
The actual provider (p.actualProvider) may not exist, IIUC. So, the lazy provider reports as "always enabled" to the credentialprovider and when it is used, the lazy provider creates an actual provider. *IIUC*

therc replied:
@rata is correct. actualProvider is only created the first time an image is pulled. Enabled() gets called much earlier than that.

therc replied:
I added a comment explaining why we always return true.

justinsb commented:
A few code nits/questions, but also:

How does cross-region ECR work in terms of authentication?  Are the IAM instance credentials for us-east-1 valid against a us-west ECR?

Do we now construct an ECR provider against every region?  I don't quite understand how the laziness works...

therc commented:
@justinsb in my testing, it looks like my IAM instance credentials in us-east-1 were enough to invoke GetAuthorizationToken in us-west.

We only create empty lazy providers in every region. This layered approach is a compromise between two extremes:

1. Create a real ecrProvider in every region, which obtains and refreshes credentials on a regular basis. This is wasteful for every node to do, especially for the vast majority of people who'll tend to use a local registry. In addition to being more likely to trigger throttling, this might cause timeouts in regions where are there no registries (and there might be none for a long time).
1. Get credentials for every image pulled.

The real ECR provider is only created the first time we try to pull an actual image.

Commit:
Include changes from feedback

Use constructor for ecrProvider
Rename package to "credentials" like golint requests
Don't wrap the lazy provider with a caching provider
Add immedita compile-time interface conformance checks for the interfaces
Added comments

therc commented:
Feedback addessed. I'm still on the fence on whether we should have just one lazy provider vs one for each and every region.

rata commented:
@therc: great!

rata commented:
@therc: I'm not familiar with the code at all, but I'm unsure why it's not better to move the lazy-ness to  the "credentialprovider.CachingDockerConfigProvider". I'll try to have a look next week, but don't worry about me. Thanks! :)

Commit:
Allow cross-region image pulling with AWS' ECR

This is step two. We now create long-lived, lazy ECR providers in all regions.
When first used, they will create the actual ECR providers doing the work
behind the scenes, namely talking to ECR in the region where the image lives,
rather than the one our instance is running in.

Also:

- moved the list of AWS regions out of the AWS cloudprovider and into the
credentialprovider, then exported it from there.
- improved logging

Behold, running in us-east-1:

```
aws_credentials.go:127] Creating ecrProvider for us-west-2
aws_credentials.go:63] AWS request: ecr:GetAuthorizationToken in us-west-2
aws_credentials.go:217] Adding credentials for user AWS in us-west-2
Successfully pulled image 123456789012.dkr.ecr.us-west-2.amazonaws.com/test:latest"
```

*"One small step for a pod, one giant leap for Kube-kind."*

Commit:
Add ap-northeast-2

therc commented:
@k8s-bot retest this issue: #21484

therc commented:
Rebased to pick up ap-northeast-2 from #24457

therc commented:
Ping?

rata commented:
@therc IMHO, I don't know if the laziness can be moved a layer up, or even if it makes sense. But in any case, I'm fine with this and improve later if there is room for improvement.

In any case, it doesn't really counts what I say. Probably the best is to ping @justinsb ? :)

therc commented:
I think this should be reassigned to @justinsb after the bot went crazy the other day. Rebased in the meantime.

therc commented:
@k8s-bot test this #22655

therc commented:
@k8s-bot test this issue: #22655

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -42,7 +42,7 @@ import (
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/cloudprovider"
-	"k8s.io/kubernetes/pkg/credentialprovider/aws"
+	aws_credentials "k8s.io/kubernetes/pkg/credentialprovider/aws"
```
This is a confusing alias (I would have guessed "github.com/aws/aws-sdk-go/aws/credentials", but I can't think of a better one that isn't 30 characters long).

rata replied:
aws_cred_provider?

justinsb commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -26,23 +27,40 @@ import (
 	}
 
-	glog.V(4).Infof("AWS request: %s %s", service, name)
+	glog.V(3).Infof("AWS request: %s:%s in %s", service, name, *region)
```
I think we probably should guard against the region being nil, but there's only a tiny number of non-regional calls AFAIK, and I think they are all S3 calls.

justinsb commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -59,56 +77,101 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
+			return nil
+		}
+	}
+	entry := p.actualProvider.Provide()[p.regionURL]
```
I think we should probably guard against an error meaning the region is not present in the DockerConfig map.

justinsb commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -59,56 +77,101 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
+}
+
+// Provide implements DockerConfigProvider.Provide, creating dummy credentials.
+// Client code will call Provider.LazyProvide() at image pulling time.
```
Should we remove the Provide() method from the DockerConfigProvider interface?  I find it confusing that there are now two methods on the interface - it really isn't clear (to me) what is called when.  But I guess that isn't this PR.

justinsb commented:
I do find this very confusing, so I'm tempted to attempt a refactor, but let's get this merged!  Thanks for implementing - will be very helpful to a lot of people.

therc commented:
I'll create another PR with a mini design doc, a link to it and miscellaneous remaining fixes, but yes, it has to be either confusing like this or very inefficient, given the existing design of credential providers.

justinsb commented:
@therc let's get the AWS PR backlog (of my fault) merged, and then maybe let's look at reworking the design of the credentials provider?  I'm certainly game to take a look, and there's no point doing a lot of documentation it if we're going to change it.  If/when I/we discover this is actually harder than I think and the design must stay, then more documentation is very welcome :-)
